File: src/main/java/uk/laxd/dndSimulator/event/EncounterEvent.java
package uk.laxd.dndSimulator.event;

import uk.laxd.dndSimulator.Effect;
import uk.laxd.dndSimulator.action.AttackOutcome;
import uk.laxd.dndSimulator.action.Damage;
import uk.laxd.dndSimulator.character.Character;
import uk.laxd.dndSimulator.equipment.Weapon;

/**
 * A collection of EncounterEvents can be used to describe an entire encounter
 * and may contain:
 *
 * Attacks
 *
 */
// TODO: Subclass this with DamageEvent, HealingEvent, ItemUseEvent etc
public class EncounterEvent implements Comparable<EncounterEvent> {

    // Index of the event
    private int index;

    // What happened
    EncounterEventType type;

    // With what
    Weapon weapon;

    // How much
    Damage amount;

    // Why did it happen?
    Effect effect;

    // Who did it
    Character actor;

    // To who
    Character target;

    // Did it work? (TODO: Generalise)
    AttackOutcome eventOutcome;

    public int getIndex() {
        return index;
    }

    public void setIndex(int index) {
        this.index = index;
    }

    public EncounterEventType getType() {
        return type;
    }

    public Weapon getWeapon() {
        return weapon;
    }

    public Damage getAmount() {
        return amount;
    }

    public Effect getEffect() {
        return effect;
    }

    public Character getActor() {
        return actor;
    }

    public Character getTarget() {
        return target;
    }

    public AttackOutcome getEventOutcome() {
        return eventOutcome;
    }

    @Override
    public String toString() {
        return String.format("%s attacked %s with %s dealing %s damage", actor, target, weapon, amount);
    }

    @Override
    public int compareTo(EncounterEvent encounterEvent) {
        return this.index - encounterEvent.index;
    }
}

File: app/src/main/kotlin/uk/laxd/dndSimulator/event/EncounterEvent.kt
package uk.laxd.dndSimulator.event

import uk.laxd.dndSimulator.action.AttackOutcome
import uk.laxd.dndSimulator.action.Damage
import uk.laxd.dndSimulator.character.Character
import uk.laxd.dndSimulator.dice.RollResult
import uk.laxd.dndSimulator.feature.Effect

/**
 * A collection of EncounterEvents can be used to describe an entire encounter
 * and may contain:
 *
 * - Melee attacks
 * - Spells
 * - Movement
 * - Generic events (Encounter start, turn start, deaths, etc)
 *
 */
// TODO: Subclass this with DamageEvent, HealingEvent, ItemUseEvent etc
open class EncounterEvent(
    val actor: Character?,
    val type: EncounterEventType,
    val string: String
) {
    override fun toString() = string
}

class GeneralEncounterEvent(actor: Character, type: EncounterEventType) : EncounterEvent(actor, type, "$type event")

class TurnStartEvent(actor: Character) : EncounterEvent(actor, EncounterEventType.TURN_START, "$actor turn started")

class TurnEndEvent(actor: Character) : EncounterEvent(actor, EncounterEventType.TURN_END, "$actor turn finished")

class RoundStartEvent : EncounterEvent(null, EncounterEventType.ROUND_START, "New round started")

class EncounterStartEvent : EncounterEvent(null, EncounterEventType.ENCOUNTER_START, "New encounter started")

class DeathEvent(character: Character, killedBy: Character) : EncounterEvent(character, EncounterEventType.DEATH, "$killedBy killed $character!")

class EncounterFinishedEvent(val winningTeam : String) : EncounterEvent(null, EncounterEventType.ENCOUNTER_FINISH, "$winningTeam won!")

class MeleeAttackEventOnly(character: Character, val target: Character, val rollResult: RollResult, val outcome: AttackOutcome) :
    EncounterEvent(character, EncounterEventType.MELEE_ATTACK, "$character attacked $target - $rollResult - $outcome")

class DamageEvent(character: Character, val target: Character, val damage: Damage, val effect: Effect) :
        EncounterEvent(character, EncounterEventType.DAMAGE, "$target took $damage damage from $character's $effect")
